2022/src/day16.scala

```diff
@@ -98,36 +98,39 @@ def bestPath(map: RoomsInfo, start: Id, valves: Set[Id], timeAllowed: Int): Int
   // we limit our options by only considering the still-closed valves
   // and `valves` has already culled any room with a flow value of 0 -- no point in considering these rooms!
 
-  val valvesArray = valves.toArray
-  val valveCount = valvesArray.length
-  val valvesIndicesLeft = Array.fill[Boolean](valveCount)(true)
-  val roomsByIndices = Array.tabulate(valveCount)(i => map.rooms(valvesArray(i)))
-
-  def recurse(current: Id, timeLeft: Int, totalValue: Int): Int =
+  val valvesLookup = IArray.from(valves)
+  val valveCount = valvesLookup.size
+  val _activeValveIndices = Array.fill[Boolean](valveCount + 1)(true) // add an extra valve for the initial state
+  def valveIndexLeft(i: Int) = _activeValveIndices(i)
+  def withoutValve(i: Int)(f: => Int) =
+    try
+      _activeValveIndices(i) = false
+      f
+    finally _activeValveIndices(i) = true
+  val roomsByIndices = IArray.tabulate(valveCount)(i => map.rooms(valvesLookup(i)))
+
+  def recurse(hiddenValve: Int, current: Id, timeLeft: Int, totalValue: Int): Int = withoutValve(hiddenValve):
     // recursively consider all plausible options
     // we are finished when we no longer have time to reach another valve or all valves are open
     val routesOfCurrent = map.routes(current)
     var bestValue = totalValue
-    for index <- 0 until valveCount do
-      if valvesIndicesLeft(index) then
-        val id = valvesArray(index)
-        val distance = routesOfCurrent(id)
-        // how much time is left after we traverse there and open the valve?
-        val t = timeLeft - distance - 1
-        // if `t` is zero or less this option can be skipped
-        if t > 0 then
-          // the value of choosing a particular valve (over the life of our simulation)
-          // is its flow rate multiplied by the time remaining after opening it
-          val value = roomsByIndices(index).flow * t
-          valvesIndicesLeft(index) = false
-          val recValue = recurse(id, t, totalValue + value)
-          valvesIndicesLeft(index) = true
-          if recValue > bestValue then
-            bestValue = recValue
+    for index <- valvesLookup.indices if valveIndexLeft(index) do
```

Comment:
That's not nearly as good. Even `indices` is not great. Putting the `if` inside the comprehension creates a dreadful `withFilter`.

Reply:
isn't indices just sugar for `0 to foo.length`?

Reply:
There's just enough indirection that the Scala.js cannot stack-allocate the `Range` anymore, and therefore does not turn the `foreach` into a `while` loop.